Treadmill elevation gain

cagnulein opened the pull request:
@victorypoint this is the branch in order to test positive and negative inclination gain to zwift.
It will build automatically a new build for windows.
Let me know

victorypoint commented:
@cagnulein, here are some test results and findings regarding treadmill elevation gain in Zwift. The tests illustrate the basic problem with treadmill elevation gain recorded in Zwift FIT files in running mode. Specifically, Zwift only records 0 and positive elevation values; negative elevation/incline values are ignored and are instead recorded as 0 elevation gain. This is either a bug in Zwift or intentionally done for running. Essentially, Zwift assumes only 0 and positive incline will be used for treadmills.

1. Zwift bike mode - We know that in bike mode, Zwift records altitude, derived from the digital elevation model of the virtual Zwift course, to it's FIT file. Viewing the elevation graph of an example FIT file in Strava looks like this:

![bike-Screenshot_20230219-233118](https://user-images.githubusercontent.com/63697253/220013718-44f081f0-13f6-4083-b195-3eb7097a9da8.png)

3. Zwift running mode with auto-inclination - When Zwift is setup to use auto-inclination in running mode, Zwift ignores the negative elevation values and replaces them with 0 in it's FIT file as noted above. Viewing the elevation graph of an test FIT file in Strava looks like this:

![zwift-Screenshot_20230219-231347](https://user-images.githubusercontent.com/63697253/220013790-3b189b38-2ce5-4ebf-83b7-05b94e4f7058.png)

QZ on the other hand records positive and negative elevation gain to its FIT file.

Using a Windows utility, Fit File Repair Tool (https://www.fitfilerepairtool.info/), a MS Access database app, we can replace the incorrect altitude records in the Zwift FIT file with correct values from the QZ FIT file while maintaining proper event timestamps and sequence. Viewing the elevation graph of the corrected FIT file in Strava looks like this:

![qz-Screenshot_20230219-232044](https://user-images.githubusercontent.com/63697253/220013839-81f3ae9a-df39-448f-a9ac-2d4940477adb.png)

4. Zwift running mode with only positive manual inclination - The idea here is if Zwift only records 0 and positive elevation gain, then FIT files from both Zwift and QZ should look identical as they are both correctly recording positive elevation gain. Viewing the elevation graph of an test FIT file, using only treadmill positive incline, in Strava looks like this:

![zwift-Screenshot_20230217-151248](https://user-images.githubusercontent.com/63697253/220013902-921990bb-ff24-4ff3-b369-d8b52ede27d9.png)

And the elevation graph of the QZ FIT file for the same test looks like this in Strava:

![qz-Screenshot_20230217-151406](https://user-images.githubusercontent.com/63697253/220013924-45fdf832-393e-410c-8b9d-fb074e3c7411.png)

This shows that both Zwift and QZ are calculating elevation gain the same for positive treadmill incline. Next we will compare elevation graphs for positive and negative incline.

5. Zwift running mode with positive and negative manual inclination - As expected, when treadmill incline is adjusted to negative values (e.g. -1, 2, 3%), instead of recording descending or negative elevation gain, Zwift again ignores the negative elevation values and replaces them with 0 in it's FIT file. Viewing the elevation graph of an test FIT file in Strava looks like this:

![zwift-Screenshot_20230217-134212](https://user-images.githubusercontent.com/63697253/220014024-1aa5a5f9-5362-4c96-8abc-dab79fe7b6c5.png)

Correcting the Zwift FIT file with Fit File Repair Tool, the corrected elevation graph in Strava looks like this:

![merge-Screenshot_20230217-144843](https://user-images.githubusercontent.com/63697253/220014055-533bc30a-af94-4014-aceb-3ac4bfa60d2b.png)

Conclusions and recommendations:

- We find that corrected Zwift FIT files work in Strava and show all the same statistics - the Zwift course or event as title, a map, distance, time, calories, heart rate, cadence, etc, but also the correct elevation graph and elevation gain. QZ power data can also be merged with a Zwift FIT file to provide power-based analysis.

- Can Zwift FIT files be intercepted and corrected before they are uploaded to Strava? After testing this scenario, I found that Zwift FIT files can't be replaced or overwritten before upload as the file is held open by Zwift. The files could however be replaced once Zwift releases the open lock on them, but would need to be manually uploaded to Strava.

- Could QZ be used to automatically or manually upload Zwift FIT activities after completion? We would need to determine if QZ can perform the database manipulation required that Fit File Repair Tool currently does. That is, replace Zwift FIT elevation values with QZ values, but also obtain positional data (latitude and longitude) from Zwift. Without positional data, QZ FIT files do not display map and elevation data in Strava.

cagnulein commented:
@victorypoint first of all, are you using the executable from this branch? here i added the positive and negative elevation to zwift by bluetooth.

> Can Zwift FIT files be intercepted and corrected before they are uploaded to Strava?

no, it's impossible. 

> Could QZ be used to automatically or manually upload Zwift FIT activities after completion? 

you can press stop on qz to automatically upload the fit file as usual.

> We would need to determine if QZ can perform the database manipulation required that Fit File Repair Tool currently does. That is, replace Zwift FIT elevation values with QZ values, but also obtain positional data (latitude and longitude) from Zwift. Without positional data, QZ FIT files do not display map and elevation data in Strava.

I think we're exiting from the scope of QZ here :D I mean probably the best way is to create a web application that merge the fit file on strava like https://github.com/strautomator

I already told with the developer in the past and he was very frindly

victorypoint commented:
> @victorypoint first of all, are you using the executable from this branch? here i added the positive and negative elevation to zwift by bluetooth.

No but I am excited to test it out. Where can I get it and what does it do regarding Zwift and QZ FIT files?

> I think we're exiting from the scope of QZ here :D I mean probably the best way is to create a web application that merge the fit file on strava like https://github.com/strautomator
> 
> I already told with the developer in the past and he was very frindly

This is the first I've heard of Strautomator. It looks very interesting!

cagnulein commented:
@victorypoint here https://github.com/cagnulein/qdomyos-zwift/suites/10998452408/artifacts/557702215

victorypoint commented:
@cagnulein, just need to confirm testing parameters for this build. You're sending QZ calculated elevation to Zwift over Bluetooth only? So don't test with Dircon? With QZ virtual device and Bluetooth on, we get 3 Bluetooth virtual bridge options. Which ones to use?

cagnulein commented:
yes i'm sending calculated elevation to zwift by bluetooth and dircon both.
use the device that you want, should be the same for qz side, of course
zwift could do something different with bluetooth or wifi, who knows! :)

Il giorno lun 20 feb 2023 alle 17:10 Al Udell ***@***.***> ha
scritto:

> @cagnulein <https://github.com/cagnulein>, just need to confirm testing
> parameters for this build. You're sending QZ calculated elevation to Zwift
> over Bluetooth only? So don't test with Dircon? With QZ virtual device and
> Bluetooth on, we get 3 Bluetooth virtual bridge options. Which ones to use?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cagnulein/qdomyos-zwift/pull/1280#issuecomment-1437254150>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALYWFYKFZ7OZHT6SLTEX3WYOJN5ANCNFSM6AAAAAAU45RE6Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud

victorypoint commented:
@cagnulein, okay I ran into a snag on the first manual incline test. Using new QZ build, fake treadmill, dircon to Zwift. Using negative incline in QZ shows no change in elevation gain (previous QZ versions show decreasing elevation gain). However, using positive incline shows increasing elevation gain which is to be expected.

![negative-incline](https://user-images.githubusercontent.com/63697253/220203559-773c695a-768f-4600-8743-a665d1052958.jpg)

![positive-incline](https://user-images.githubusercontent.com/63697253/220203572-55b31abb-8067-4e38-aa13-78db752bb2a5.jpg)

cagnulein commented:
yes it's by design because i quickly made a custom proprieties in order to send them to zwift. does zwift changes the behavior with this build to strava so?

victorypoint commented:
> yes it's by design because i quickly made a custom proprieties in order to send them to zwift. does zwift changes the behavior with this build to strava so?

It appears that Zwift still zeros negative elevation gain however the Strava graph shows 1 little dip in the middle.

![zwift-Screenshot_20230220-175623](https://user-images.githubusercontent.com/63697253/220204991-d7270e81-c525-44d8-9d22-7a576d612f17.png)

Strava elevation graph for the merged FIT file (using FIT File Repair Tool) shows QZ is ignoring negative elevation gain as well.

![merge-Screenshot_20230220-180047](https://user-images.githubusercontent.com/63697253/220205107-e0a0355d-466e-455e-8d40-4d916d283eee.png)

[manual-incline-test.zip](https://github.com/cagnulein/qdomyos-zwift/files/10788039/manual-incline-test.zip)

victorypoint commented:
Manual inclines I used:

	Mon Feb 20 17:39:19 2023 1676929159951 Debug: treadmill.cpp virtual void treadmill::changeSpeed(double) changeSpeed 10 true 1 0
	Mon Feb 20 17:39:33 2023 1676929173445 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -3 true 1 0
	Mon Feb 20 17:41:19 2023 1676929279186 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -3.5 true 1 0
	Mon Feb 20 17:41:20 2023 1676929280344 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -4 true 1 0
	Mon Feb 20 17:41:22 2023 1676929282177 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -4.5 true 1 0
	Mon Feb 20 17:41:23 2023 1676929283171 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -5 true 1 0
	Mon Feb 20 17:41:24 2023 1676929284189 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -5.5 true 1 0
	Mon Feb 20 17:41:25 2023 1676929285339 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -6 true 1 0
	Mon Feb 20 17:41:27 2023 1676929287189 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -6.5 true 1 0
	Mon Feb 20 17:41:28 2023 1676929288779 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -7 true 1 0
	Mon Feb 20 17:41:30 2023 1676929290580 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -7.5 true 1 0
	Mon Feb 20 17:41:32 2023 1676929292194 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -8 true 1 0
	Mon Feb 20 17:41:34 2023 1676929294185 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -8.5 true 1 0
	Mon Feb 20 17:41:36 2023 1676929296192 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -9 true 1 0
	Mon Feb 20 17:41:38 2023 1676929298235 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -9.5 true 1 0
	Mon Feb 20 17:41:40 2023 1676929300205 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -10 true 1 0
	Mon Feb 20 17:43:27 2023 1676929407740 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination 4 true 1 0
	Mon Feb 20 17:43:29 2023 1676929409731 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination 4 true 1 0
	Mon Feb 20 17:45:27 2023 1676929527218 Debug: treadmill.cpp virtual void treadmill::changeInclination(double, double) changeInclination -3 true 1 0

cagnulein commented:
thanks @victorypoint so zwift just ignores them, so even this branch is useless.
There are only 2 way to achieve this:
1. zwift should patch their code base
2. stravautomator new feature with a merge of qz and zwift file (i guess it's highly doable)

I'm gonna open a new ticket to his repository. I'm tagging you

victorypoint commented:
@cagnulein, just curious if your able to obtain Zwift latitude and longitude data during a Zwift run for inclusion in QZ FIT file?

cagnulein commented:
no I can't, zwift doesn't send it. The only way to do so is with a man in the middle technique or with a 2nd guest account that "observe" the main account. Very very tricky so I will not proceed with any of these 2 options, they are too much dipendent from the zwift side
